### ISSUE TITLE
[XF] MasterDetailPage navigation within FormsAppCompatActivity

### ISSUE THREAD

**Issue by Adam--** [PrismLibrary/Prism#399]
When a MasterDetailPage is loaded with the Detail set to a ContentPage wrapped in a NavigationPage (`MasterDetailPage/NavigationPage/ContentPage`) and navigation is performed from the MasterDetailPage's view model to some other content page (`NavigationService.Navigate("SomeOtherContentPage")`)  it is expected that the toolbar will still be shown.

For an Android MainActivity that derives from FormsApplicationActivity this is the case; however, if it derives from FormsAppCompatActivity, the toolbar is not shown. If navigation is called explicitly with a NavigationPage (`NavigationService.Navigate("NavigationPage/SomeOtherContentPage")`), the toolbar is shown.

This behavior can be observed in the current Sandbox/Xamarin/HelloWorld solution.
![masterdetailnavigation appcompat bug](https://cloud.githubusercontent.com/assets/2179171/12362392/126a6c60-bb91-11e5-927c-87eb11310f4a.gif)

App.cs:
```
    public class App : PrismApplication
    {
        protected override void OnInitialized()
        {
            NavigationService.Navigate("MyMasterDetail/MyNavigationPage/ViewA");
        }
...
```

MyMasterDetailViewModel.cs:
```
    public class MyMasterDetailViewModel : ViewModelBase
    {
        INavigationService _navigationService;

        public DelegateCommand<string> NavigateCommand { get; set; }
        public MyMasterDetailViewModel(INavigationService navigationService)
        {
            _navigationService = navigationService;
            NavigateCommand = new DelegateCommand<string>(Navigate);
        }

        private void Navigate(string name)
        {
            _navigationService.Navigate(name);
        }
```

MyMasterDetail.xaml:
```
...
  <MasterDetailPage.Master>
    <ContentPage Title="My MatserDetailPage">
      <StackLayout>
        <Button Text="ViewA" Command="{Binding NavigateCommand}" CommandParameter="ViewA?id=A" />
        <Button Text="ViewB" Command="{Binding NavigateCommand}" CommandParameter="ViewB?id=B" />
        <Button Text="ViewC" Command="{Binding NavigateCommand}" CommandParameter="ViewC?id=C" />
      </StackLayout>
    </ContentPage>
  </MasterDetailPage.Master>
....
```

**Comment by brianlagunas** [comment]
The MasterDetail toolbar functions differently between FormsApplicationActivity and FormsAppCompatActivity, but that is not the root of your navigation issue.

The reason the toolbar is showing initially is because you have the ContentPage wrapped in a NavigationPage.  Once you navigate by clicking the "ViewA" button, you have now navigated to a new view.  Since you made the navigation call from the MasterDetailViewModel, this replaces the current Detail with the new target View, which is not wrapped in a NavigationPage, and removes the NavigationPage/ViewA completely.  So hence, your toolbar is removed.

You have to rememebrt hat navigation is always based on the calling page.  So if you want to maintain the same NavigationPage in the MasterDetai.Detail, then you have to navigate off of the NavigationPage.Current instead, and not the MasterPage itself.

So you would want to do something like this:
```
<MasterDetailPage.Master>
    <ContentPage Title="My MatserDetailPage">
      <StackLayout>
        <Button Text="ViewA" BindingContext="{x:Reference masterDetail}" Command="{Binding Detail.CurrentPage.BindingContext.NavigateCommand}" CommandParameter="ViewA?id=A" />
        <Button Text="ViewB" BindingContext="{x:Reference masterDetail }" Command="{Binding Detail.CurrentPage.BindingContext.NavigateCommand}" CommandParameter="ViewB?id=B" />
        <Button Text="ViewC" BindingContext="{x:Reference masterDetail }" Command="{Binding Detail.CurrentPage.BindingContext.NavigateCommand}" CommandParameter="ViewC?id=C" />
      </StackLayout>
    </ContentPage>
  </MasterDetailPage.Master>
```

**Comment by Adam--** [comment]
>The MasterDetail toolbar functions differently between FormsApplicationActivity and FormsAppCompatActivity, but that is not the root of your navigation issue.

Sorry, thought that when navigating from the MasterDetail it was checking to see if the MasterDetail's Detail was a NavigationPage and if so maintaining it and removing the current page from its stack and adding a new one. My buddy demoed what looked to be that behavior with a FormsApplicationActivity and it appeared to be the case since the toolbar seemed to be maintained. I'll ask him to post an example.

To get the expected master-detail behavior described, I would expect the code would need to be:

```
  <MasterDetailPage.Master>
    <ContentPage Title="My MatserDetailPage">
      <StackLayout>
        <Button Text="ViewA" Command="{Binding NavigateCommand}" CommandParameter="MyNavigationPage/ViewA?id=A" />
        <Button Text="ViewB" Command="{Binding NavigateCommand}" CommandParameter="MyNavigationPage/ViewB?id=B" />
        <Button Text="ViewC" Command="{Binding NavigateCommand}" CommandParameter="MyNavigationPage/ViewC?id=C" />
      </StackLayout>
    </ContentPage>
  </MasterDetailPage.Master>
```

I know it's not meant to be a sample, but maybe the sandbox HelloWorld should be updated or maybe a navigation sample should be provided.

**Comment by brianlagunas** [comment]
I'm curious why you would expect that behavior with your snippet?  If it were to behave the way you were expecting, how would you replace the Detail with a new NavigationPage?  Also, what would be the point of removing the previous page form the NavigationPage and adding a new one?  You would gain no benefit since you would not have the navigation history in the NavigationPage.  So it appears you would be doing this for the sole purpose of keeping the toolbar.  Would that be an accurate statement?

**Comment by Adam--** [comment]
For the first code snippet with `CommandParameter="ViewA?id=A"`. I only expected that behavior due to the example of the FormsApplicationActivity shown to me and assumed that behavior would be maintained to FormsAppCompatActivity. Thinking about it though, it doesn't make any sense to expect that behavior in either case. Although, I don't understand why the toolbar is still shown in FormsApplicationActivity when navigating to ContentPage without a NavigationPage.

For the second code snippet with `CommandParameter="MyNavigationPage/ViewA?id=A"` Yes I would do this to keep the toolbar. You wouldn't want to remove visual indication that the navigation drawer is available.

**Comment by craigblazakis** [comment]
The following animation illustrates what I observed using code that is configured as described by Adam in his original post.  When navigating from the MasterDetail to a content page that is not wrapped in a NavigationPage, the toolbar is still displayed.

![loop](https://cloud.githubusercontent.com/assets/16725744/12365463/043e246c-bba3-11e5-81f9-a0ef69c1312d.gif)

**Comment by brianlagunas** [comment]
Yes, that is the expected behavior when using FormsAppCompatActivity.  The MasterDetailPage.Detail page will only display a navigation bar if it is an instance of NavigationPage.

So if all you want is to have the toolbar show for everypage, but don't want any of the navigation history, then just place all the page in a NavigationPage like in your above smippet.  Just realize, this is a new NavigationPage instance on every navigation.

**Comment by brianlagunas** [comment]
The Sandbox project is my playground for playing with features.  When the navigation functionality stabilizes I will create a MasterDetail sample.

**Comment by brianlagunas** [comment]
Just so you know, I am trying to think of how Prism can make this more intuitive and easier to use.  Basically hide the platform limitations/annoyances from the user.  I am going to research using attached properties to control the NavigationPage scenarios for MasterDetailPages.

**Comment by brianlagunas** [comment]
**Correction**:  If your snippet is:
```
<StackLayout>
        <Button Text="ViewA" Command="{Binding NavigateCommand}" CommandParameter="MyNavigationPage/ViewA?id=A" />
        <Button Text="ViewB" Command="{Binding NavigateCommand}" CommandParameter="MyNavigationPage/ViewB?id=B" />
        <Button Text="ViewC" Command="{Binding NavigateCommand}" CommandParameter="MyNavigationPage/ViewC?id=C" />
      </StackLayout>
```
Then the NavigationPage is reused, and we are removing the previous page, and adding the new one.  I got confused.  Anytime the Detail is the same as the target page, it is reused.

I guess my question to you would be, should we modify the NavigationPage logic to always add pages to the NavigationPage stack, or remove them like we do now.

**Comment by brianlagunas** [comment]
Okay, the more I look at this the more I think the ProcessNavigationForNavigationPage logic needs to be rethought.  I think some improvements can be made there.

Some decisions need to be made regarding navigating from a NavigationPage.  Right now the logic goes something like this:

* If NavigationPage.NavigationStack is empty, then create a new page.
* Check the first element in the NavigationPage.NavigationStack, and if that page is the same as the next, and there are more items on the stack, then PopToRoot, and continue. **(should this change)**?
* If the root page page is different, then PopToRoot, and continue - **(should this change)**

Basically, should we always clear the stack when navigating from a NavigationPage, or should we just always keep pushing onto the NavigationStack?

**Comment by powerdude** [comment]
I'd vote for not automatically clearing the stack and add a parameter to allow someone to do so.

**Comment by Adam--** [comment]
One use case is when Navigate is called from a MasterDetail VM when the detail is a NavigationPage. This logic seems correct in that case. What would be the other use cases? 

>  I am going to research using attached properties to control the NavigationPage scenarios for MasterDetailPages.
Sounds helpful.

**Comment by brianlagunas** [comment]
I actually found a bug in Xamarin.Forms which is causing the MasterDeatil button to render incorrectly when reusing the NavigationPage.

We can track it here:  https://bugzilla.xamarin.com/show_bug.cgi?id=38105
